core/state_processor.go

```diff
@@ -59,20 +60,28 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 		usedGas  = new(uint64)
 		header   = block.Header()
 		allLogs  []*types.Log
-		gp       = new(GasPool).AddGas(CalcGasLimit(block, statedb))
+		vmRunner = p.bc.NewSystemEVMRunner(block.Header(), statedb)
+		gp       = new(GasPool).AddGas(blockchain_parameters.GetBlockGasLimitOrDefault(vmRunner))
 	)
 	// Mutate the block and state according to any hard-fork specs
 	if p.config.DAOForkSupport && p.config.DAOForkBlock != nil && p.config.DAOForkBlock.Cmp(block.Number()) == 0 {
 		misc.ApplyDAOHardFork(statedb)
 	}
 
-	if random.IsRunning() {
+	// TODO(HF) use vmRunner instead of parentVMRunner (HF required)
```

Comment:
i guess... can we assume that `state_processor` is only called with block N+1 when blockchain head is N? I was not entirely sure about that.

Reply:
It's only used in `insertChain()`.  Now, I can think of two scenarios of `insertChain()`:

1. Adding a new block to the head of the chain
2. Inserting a block into a sidechain

Though since you reviewed the disable reorg PR you know better than me whether there are other cases.

For #1, we're fine.  For #2, well, we're going to be disabling reorgs and so it won't happen.  But, even without disabling reorgs, if it does happen, wouldn't using the chain head instead of the block being inserted mean that different nodes could use different states (while processing the same block) depending on the timing (e.g. a new block comes in)?  In other words, it's not like the current behavior results in different nodes agreeing with each other in the first place.  Though this is largely academic for the long-lived networks, since `IsRunning` just checks whether the contract has been deployed and registered (i.e. at some point it goes from false to true and then it stays true).

Still, I guess if we want to be super-safe we can do this as part of a hard fork.  But then do we have to keep the conditional forever?  Or do we try to sneakily remove it and pretend it was never part of a HF?

Reply:
i'll take the simplistic approach an use current state.

As you said, this only happens once over the blockchain history. And that's on migrations. And it's unlikely that you call `Process()` on a block that is not currentHead +1.

For starters, consensus engine will also use a N+1 block, then fetcher/downloader will only use N+1. Maybe a `SetHead()` or `reorg`, but i dont' think so. And even if that were the case, it is a bug with un specified behvaiour, so better to fix this here.